framework/src/utils/TransitionLayerTools.C

```diff
@@ -0,0 +1,623 @@
+    { return old_id_pair.first == end_node_id; };
+    auto isMatch2 = [end_node_id](std::pair<dof_id_type, dof_id_type> old_id_pair)
+    { return old_id_pair.second == end_node_id; };
+    auto result = std::find_if(boundary_node_assm.begin(), boundary_node_assm.end(), isMatch1);
+    bool match_first;
+    if (result == boundary_node_assm.end())
+    {
+      match_first = false;
+      result = std::find_if(boundary_node_assm.begin(), boundary_node_assm.end(), isMatch2);
+    }
+    else
+    {
+      match_first = true;
+    }
+    // If found, add the node to boundary_ordered_node_list
+    if (result != boundary_node_assm.end())
+    {
+      boundary_ordered_node_list.push_back(match_first ? (*result).second : (*result).first);
+      boundary_node_assm.erase(result);
+    }
+    // If there are still elements in boundary_node_assm and result ==
+    // boundary_node_assm.end(), this means the boundary is not a loop, the
+    // boundary_ordered_node_list is flipped and try the other direction that has not
+    // been examined yet.
+    else
+    {
+      if (isFlipped)
+      {
+        // Flipped twice; this means the boundary has at least two segments.
+        // This is invalid type #1
+        throw(unsigned short) 1;
+        return false;
+      }
+      // mark the first flip event.
+      isFlipped = true;
+      std::reverse(boundary_ordered_node_list.begin(), boundary_ordered_node_list.end());
+      // As this iteration is wasted, set the iterator backward
+      i--;
+    }
+  }
+  // If the code ever gets here, boundary_node_assm is empty.
+  // If the boundary_ordered_node_list front and back are not the same, the boundary is not a loop.
+  // This is not done inside the loop just for some potential applications in the future.
+  if (boundary_ordered_node_list.front() != boundary_ordered_node_list.back())
+  {
+    // This is invalid type #2
+    throw(unsigned short) 2;
+    return false;
+  }
+  // It the boundary is a loop, check if azimuthal angles change monotonically
+  else
+  {
+    // Utilize cross product here.
+    // If azimuthal angles change monotonically,
+    // the z components of the cross products are always negative or positive.
+    std::vector<Real> ordered_node_azi_list;
+    for (unsigned int i = 0; i < boundary_ordered_node_list.size() - 1; i++)
+    {
+      ordered_node_azi_list.push_back(
+          (*mesh.node_ptr(boundary_ordered_node_list[i]) - origin_pt)
+              .cross(*mesh.node_ptr(boundary_ordered_node_list[i + 1]) - origin_pt)(2));
+      // Use this opportunity to calculate maximum radius
+      max_node_radius = std::max((*mesh.node_ptr(boundary_ordered_node_list[i]) - origin_pt).norm(),
+                                 max_node_radius);
+    }
+    std::sort(ordered_node_azi_list.begin(), ordered_node_azi_list.end());
+    if (ordered_node_azi_list.front() * ordered_node_azi_list.back() < 0.0)
+    {
+      // This is invalid type #3
+      throw(unsigned short) 3;
```

Comment:
use real exceptions. see the throw/catch MooseExceptions  in the framework for example